tvx/builders/generator.go

```diff
@@ -0,0 +1,166 @@
+package builders
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sync"
+
+	"github.com/filecoin-project/oni/tvx/schema"
+)
+
+// Generator is a batch generator and organizer of test vectors.
+//
+// Test vector scripts are simple programs (main function). Test vector scripts
+// can delegate to the Generator to handle the execution, reporting and capture
+// of emitted test vectors into files.
+//
+// Generator supports the following CLI flags:
+//
+//  -o <directory>
+//		directory where test vector JSON files will be saved.
+//
+//  -f <regex>
+//		regex filter to select a subset of vectors to execute; matched against
+//	 	the vector's ID.
+//
+// Scripts can bundle test vectors into "groups". The generator will execute
+// each group in parallel, and will write each vector in a file:
+// <output_dir>/<group>-<sequential>.json
+type Generator struct {
+	OutputPath string
+	Filter     *regexp.Regexp
+
+	wg sync.WaitGroup
+}
+
+type MessageVectorGenItem struct {
+	Metadata *schema.Metadata
+	Func     func(*Builder)
+}
+
+func NewGenerator() *Generator {
+	// Consume CLI parameters.
+	var (
+		outputDir = flag.String("o", "", "output directory")
+		filter    = flag.String("f", "", "filter regex for test vectors to generate")
+	)
+
+	flag.Parse()
+
+	ret := new(Generator)
+
+	// Output directory is compulsory.
+	if *outputDir == "" {
```

Comment:
Hmm, I thought the idea was to output to `os.Stdout` if this was not set?

Reply:
Yup! I'm gonna implement that now. Was thinking on the spot during the call ;-)

Reply:
I like your style 😁